src/commands/publish/worker_bundle.rs

```diff
@@ -0,0 +1,113 @@
+use crate::workers::{Resource, Worker};
+use reqwest::multipart;
+use serde::Serialize;
+
+#[derive(Debug)]
+pub struct WorkerBundle {
+    name: String,
+    script_name: String,
+    script_path: String,
+    bindings: Vec<Binding>,
+}
+
+impl WorkerBundle {
+    pub fn multipart(&self) -> Result<multipart::Form, failure::Error> {
+        let mut form = multipart::Form::new();
+        let parts = self.binding_parts()?;
+        for (name, part) in parts {
+            form = form.part(name, part);
+        }
+        form = form.file(self.script_name.clone(), &self.script_path)?;
+        Ok(form)
+    }
+
+    fn binding_parts(&self) -> Result<Vec<(String, multipart::Part)>, failure::Error> {
+        let mut parts: Vec<(String, multipart::Part)> = Vec::new();
+        let meta_part = self.binding_meta_part()?;
+        parts.push(("metadata".to_string(), meta_part));
+        for binding in &self.bindings {
+            match binding {
+                Binding::wasm_module(ref wasm) => {
+                    let part = multipart::Part::file(wasm.path.to_string())?;
+                    parts.push((wasm.name.clone(), part));
+                }
+                Binding::kv_namespace(_) => {
+                    // kv bindings don't add their own part to the multipart form
+                }
+            }
+        }
+        Ok(parts)
+    }
+
+    fn binding_meta_part(&self) -> Result<multipart::Part, failure::Error> {
+        let metadata = MetaData {
+            body_part: self.script_name.clone(),
+            bindings: &self.bindings,
+        };
+        Ok(multipart::Part::bytes(serde_json::to_vec(&metadata)?)
+            .file_name("metadata.json")
+            .mime_str("application/json")?)
+    }
+}
+
+impl From<Worker> for WorkerBundle {
+    fn from(worker: Worker) -> WorkerBundle {
+        WorkerBundle {
+            name: worker.name,
+            script_name: worker.script.name,
+            script_path: worker.script.path,
+            bindings: worker
+                .resources
+                .into_iter()
+                .map(|r| Binding::from(r))
+                .collect(),
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+pub struct MetaData<'a> {
+    body_part: String,
+    bindings: &'a [Binding],
+}
+
+// setting the serde tag to type makes it so the "type" key in the resulting json object is the
+// same as the enum member name.
+#[derive(Serialize, Debug)]
+#[serde(tag = "type")]
+pub enum Binding {
+    #[allow(non_camel_case_types)]
+    wasm_module(WasmBinding),
+    #[allow(non_camel_case_types)]
+    kv_namespace(KVBinding),
+}
+
+impl From<Resource> for Binding {
+    fn from(resource: Resource) -> Binding {
+        match resource {
+            Resource::WasmModule(wasm) => Binding::wasm_module(WasmBinding {
+                path: wasm.path.clone(),
```

Comment:
are the clones necessary?

Reply:
Just the one on the first `wasm.binding`, since it's used twice. Removing the others. Can't take references here since `resource` is consumed by this function. I can change this to move away from the `From` trait to not consume if you think that would be cleaner.

Reply:
nah, per in air convo the one clone is totally fine :)